File: src/token.py
from enum import Enum
from typing import Dict, Any, Tuple


class TokenType(Enum):
    # Single-character tokens.
    LEFT_PAREN = '('
    RIGHT_PAREN = ')'
    LEFT_BRACE = '{'
    RIGHT_BRACE = '}'
    COMMA = ','
    SEMICOLON = ';'
    SLASH = '/'
    BACKSLASH = '\\'
    STAR = '*'
    UNDERSCORE = '_'
    PERCENT = '%'
    POWER = '^'

    # One or two+ character tokens.
    BANG = '!'
    BANG_EQUAL = '!='
    EQUAL = '='
    EQUAL_EQUAL = '=='
    GREATER = '>'
    GREATER_EQUAL = '>='
    LESS = '<'
    LESS_EQUAL = '<='
    MINUS = '-'
    MINUS_MINUS = '--'
    PLUS = '+'
    PLUS_PLUS = '++'
    DOT = '.'
    ELLIPSIS = '...'
    QUESTION_MARK = '?'
    COLON = ':'
    QUESTION_MARK_COLON = '?:'
    DOUBLE_QUESTION_MARK = '??'
    PIPE = '|>'

    # Keywords.
    AND = 'and'
    CLASS = 'class'
    FALSE = 'false'
    FUN = 'fun'
    FOR = 'for'
    NIL = 'nil'
    OR = 'or'
    PRINT = 'print'
    RETURN = 'return'
    SUPER = 'super'
    SELF = 'self'
    TRUE = 'true'
    VAR = 'var'
    WITH = 'with'
    CONCAT = 'concat'
    MODULE = 'module'
    TRAIT = 'trait'
    EXTENDS = 'extends'
    BREAK = 'break'
    CONTINUE = 'continue'
    USE = 'use'

    # String starters.
    SINGLE_QUOTE = "'"
    DOUBLE_QUOTE = '"'

    # New line.
    NEWLINE = '\n'

    # Space.
    SPACE = ' '
    TAB = '\t'

    # String terminator
    NULL_CHAR = '\0'

    # EOF
    EOF = ''

    # Comment
    COMMENT = '#'
    BLOCK_COMMENT_START = '<#'
    BLOCK_COMMENT_END = '#>'

    # Literals
    IDENTIFIER = 'IDENTIFIER'
    STRING = 'STRING'
    NUMBER = 'NUMBER'


_keywords: Tuple = (
    'and', 'class', 'false', 'fun', 'for', 'nil', 'or', 'print', 'return', 'super', 'self', 'true', 'var', 'with',
    'concat', 'module', 'trait', 'extends', 'break', 'continue', 'use'
)

KEYWORDS: Dict[str, TokenType] = {key: TokenType(key) for key in _keywords}

SINGLE_CHARACTER_TOKENS: Tuple = ('(', ')', '{', '}', ',', ';', '/', '\\', '*', '_')

ONE_OR_MORE_CHARACTER_TOKENS: Tuple = (
    '!', '!=', '=', '==', '>', '>=', '<', '<=', '-', '--', '+',
    '++', '.', '...', '?', ':', '?:', '??', '|>', '#', '<#', '#>'
)

COMMENT_TOKENS: Tuple = ('#', '<#', '#>')

WHITESPACE: Tuple = (' ', '\t', '\r')

STRING_STARTERS: Tuple = ("'", '"')


class Token:
    def __init__(self, token_type: TokenType, lexeme: str, literal: Any, line: int):
        self.type = token_type
        self.lexeme = lexeme
        self.literal = literal
        self.line = line

    def __str__(self) -> str:
        return f"{self.type}: {self.lexeme}, {self.literal}, {self.line}"

    def __repr__(self) -> str:
        properties = f"{self.type}, {self.lexeme}, {self.literal}, {self.line}"
        return f'{self.__class__.__name__}({properties})'

File: compiler/src/Token.fs
/// Token types and the `Token` record produced by `Scanner`. Mirrors the
/// token set `poc/src/token.py` defines (including the keywords reserved
/// for future mixin/trait/module support, which have no grammar yet), but
/// as an idiomatic F# discriminated union rather than a `str`-backed enum
/// -- there's no need here for TokenType members that only ever existed in
/// the Python version to double as raw-character dispatch keys (comment
/// markers, whitespace, string-quote characters, EOF-as-empty-string):
/// Scanner.fs matches on plain `char`/`string` literals for those instead.
module Iqalox.Token

type TokenType =
    // Single-character tokens.
    | LeftParen
    | RightParen
    | LeftBrace
    | RightBrace
    | LeftBracket
    | RightBracket
    | Comma
    | Semicolon
    | Slash
    | Backslash
    | Star
    | Underscore
    | Percent
    | Power

    // One- or two-plus-character tokens.
    | Bang
    | BangEqual
    | Equal
    | EqualEqual
    | Greater
    | GreaterEqual
    | Less
    | LessEqual
    | Minus
    | MinusMinus
    | Arrow
    | Plus
    | PlusPlus
    | Dot
    | Ellipsis
    | QuestionMark
    | Colon
    | QuestionMarkColon
    | DoubleQuestionMark
    | Pipe

    // Keywords.
    | And
    | Class
    | False
    | Fun
    | For
    | Nil
    | Or
    | Return
    | Super
    | Self
    | True
    | Var
    | With
    | Module
    | Trait
    | Extends
    | Break
    | Continue
    | Use
    | Mutable

    // Literals.
    | Identifier
    | String
    | Number

    | Eof

/// A token's literal value, if it has one. `NoLiteral` covers every
/// non-literal token (punctuation, keywords, identifiers, EOF).
type Literal =
    | NoLiteral
    | NumberLiteral of float
    | StringLiteral of string

type Token =
    { Type: TokenType
      Lexeme: string
      Literal: Literal
      Line: int
      /// 1-indexed column of the lexeme's first character within its line.
      Column: int }

/// Keywords, per `poc/src/token.py`'s `_keywords` tuple. `with`/`module`/
/// `trait`/`use` are reserved for future mixin/trait/module support (0.2)
/// and scan as keywords here too, even though they have no grammar yet --
/// matching `poc`'s behavior, so an identifier can't accidentally shadow
/// one before that grammar lands.
let keywords: Map<string, TokenType> =
    Map.ofList [
        "and", And
        "class", Class
        "false", False
        "fun", Fun
        "for", For
        "nil", Nil
        "or", Or
        "return", Return
        "super", Super
        "self", Self
        "true", True
        "var", Var
        "with", With
        "module", Module
        "trait", Trait
        "extends", Extends
        "break", Break
        "continue", Continue
        "use", Use
        "mut", Mutable
    ]
